cmd/controller-manager/app/controller_manager.go

```diff
@@ -293,21 +293,37 @@ func StartControllers(s *options.ControllerManagerServer,
 	recorder record.EventRecorder,
 	stop <-chan struct{}) error {
 
-	// Get available service-catalog resources
-	glog.V(5).Info("Getting available resources")
-	availableResources, err := getAvailableResources(serviceCatalogClientBuilder)
-	if err != nil {
-		return err
-	}
-
-	coreKubeconfig = rest.AddUserAgent(coreKubeconfig, controllerManagerAgentName)
-	coreClient, err := kubernetes.NewForConfig(coreKubeconfig)
-	if err != nil {
-		glog.Fatal(err)
+	// When Catalog Controller and Catalog API Server are started at the
+	// same time with API Aggregation enabled, it may take some time before
+	// Catalog registration shows up in API Server.  Re-query several times if
+	// necessary.
+	var availableResources map[schema.GroupVersionResource]bool
+	catalogResourceAvailable := false
+	const maxAttempts = 4
+	for attempt := 1; attempt <= maxAttempts && !catalogResourceAvailable; attempt++ {
+		glog.V(5).Info("Getting available resources")
+		availableResources, err := getAvailableResources(serviceCatalogClientBuilder)
+		if err != nil {
+			return err
+		}
+		catalogResourceAvailable = availableResources[catalogGVR]
+		if !catalogResourceAvailable {
+			if attempt+1 <= maxAttempts {
+				glog.V(5).Info("Catalog resources are not available from API Server, will reattempt")
+				time.Sleep(10 * time.Second)
+			} else {
+				glog.V(5).Infof("Failed to get catalog resources from API Server after %d attempts.", attempt)
```

Comment:
I don't like delaying returning from the function until later, when the context has been somewhat lost. Let's just do the return here.

I would also prefer to just break out of the for loop when the catalog resource is available rather than carrying around the `catalogResourceAvailable` variable.

Reply:
refined per @staebler's review